### PR TITLE
Enhanced component manager ui

### PR DESCRIPTION
**Pull request by elfilibusterismo**
Rename activities to match sketchware activity names

Fix custom components listener not showing when changing component name.

### CODE REVIEW

**Comment by elfilibusterismo** [comment]
Screenshot_2023-08-26-12-33-09-19_423e58f25c30a02453fa5d8c82fc37f2
Here's the preview

Now you can change the component to behave like original components 
ex. FirebaseDB

**Comment by elfilibusterismo** [comment]
And of course before doing the fix.
I already studied how Component works and behave so...

File: app/src/main/java/mod/elfilibustero/sketch/editor/component/CollapsibleCustomComponentLayout.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+import java.util.List;
+
+import com.besome.sketch.lib.ui.CollapsibleButton;
+import com.besome.sketch.lib.base.CollapsibleLayout;
+import com.sketchware.remod.R;
+
+import mod.hey.studios.util.Helper;
+
+public class CollapsibleCustomComponentLayout extends CollapsibleLayout<CollapsibleButton> {
```
Comment (JavkhlanK) [inline app/src/main/java/mod/elfilibustero/sketch/editor/component/CollapsibleCustomComponentLayout.java:17]: thank you for not copy-pasting code 🙏🙏🙏

File: app/src/main/java/mod/elfilibustero/sketch/lib/ui/SketchFilePickerDialog.java [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+            }
+        });
+        super.b("Select", _v -> {
+            if (filePath != null && !filePath.isDirectory()) {
+                if (onFileSelectedListener != null) {
+                    onFileSelectedListener.onFileSelected(this, filePath);
+                }
+            }
+        });
+        super.a("Cancel", _v -> dismiss());
```
Comment (JavkhlanK) [inline app/src/main/java/mod/elfilibustero/sketch/lib/ui/SketchFilePickerDialog.java:140]: pretty sure you can use `Helper#getDialogDismissListener(Dialog)` instead

File: app/src/main/java/mod/hilal/saif/components/ComponentsHandler.java [hunk 3]
```diff
@@ -386,12 +386,12 @@ public static String var(int id) {
                         int componentIdInteger = Integer.parseInt((String) componentId);
 
                         if (componentIdInteger == id) {
-                            Object componentVarName = component.get("varName");
+                            Object componentVarName = component.get("typeName");
 
                             if (componentVarName instanceof String) {
```
Comment (JavkhlanK) [inline app/src/main/java/mod/hilal/saif/components/ComponentsHandler.java:391]: maybe change the local variable name too while at it

File: app/src/main/java/mod/hilal/saif/components/ComponentsHandler.java [hunk 4]
```diff
@@ -455,7 +455,7 @@ public static String extraVar(String name, String code, String varName) {
         for (int i = 0; i < cachedCustomComponents.size(); i++) {
             HashMap<String, Object> component = cachedCustomComponents.get(i);
             if (component != null) {
-                Object componentName = component.get("name");
+                Object componentName = component.get("typeName");
 
                 if (componentName instanceof String) {
```
Comment (JavkhlanK) [inline app/src/main/java/mod/hilal/saif/components/ComponentsHandler.java:460]: same renaming thing

File: app/src/main/java/mod/hilal/saif/components/ComponentsHandler.java [hunk 5]
```diff
@@ -488,7 +488,7 @@ public static String defineExtraVar(String name, String varName) {
...
```
Comment (JavkhlanK) [inline app/src/main/java/mod/hilal/saif/components/ComponentsHandler.java:493]: same renaming thing

File: app/src/main/res/values/strings.xml [hunk 6]
```diff
@@ -1357,6 +1360,10 @@ You must enter the API key to operate the map normally."</string>
     <string name="design_tab_title_component">Component</string>
     <string name="design_tab_title_event">Event</string>
     <string name="design_tab_title_view">View</string>
+    <string name="developer_tool_component_description_guide_add_component">"If you add a component here
+ You can use it in Logic Editor"</string>
+    <string name="developer_tool_component_message_export">Are you sure you want to export %s ?</string>
```
Comment (JavkhlanK) [inline app/src/main/res/values/strings.xml:1365]: i don't think there should be a space before the `?`, that makes it look like engrish

File: app/src/main/res/values/strings.xml [hunk 7]
```diff
@@ -1357,6 +1360,10 @@ You must enter the API key to operate the map normally."</string>
...
+    <string name="developer_tool_component_success_message_export">Successfully exported component to:\n%s</string>
```
Comment (JavkhlanK) [inline app/src/main/res/values/strings.xml:1366]: i think the prefix should be `developer_tools` since it's never mentioned as singular tool

File: app/src/main/AndroidManifest.xml
Comment (JavkhlanK) [inline app/src/main/AndroidManifest.xml:1]: this commit on its own won't compile. could you please change it so that it does? just rename the old ComponentsMaker and ComponentsMakerCreator classes to the new names alongside changing their names in the manifest in one commit, that's all that's needed

File: app/src/main/java/mod/hilal/saif/activities/tools/AddCustomComponentActivity.java [hunk 8]
```diff
@@ -0,0 +1,258 @@
...
+
+    private void selectComponentToImport(String path) {
+        String text = FileUtil.readFile(path);
+        if (text.isEmpty() || text.equals("[]")) {
+            SketchwareUtil.toastError(Helper.getResString(R.string.common_message_selected_file_empty));
+            return;
+        }
+
+        ArrayList<HashMap<String, Object>> list = new Gson().fromJson(text, Helper.TYPE_MAP_LIST);
+        if (list.isEmpty() || !ComponentsHandler.isValidComponentList(list)) {
```
Comment (JavkhlanK) [inline app/src/main/java/mod/hilal/saif/activities/tools/AddCustomComponentActivity.java:213]: this new method is already used here, but you really create it later. please fix the order, e.g. by just creating a new commit before this saying something like _add new component validation methods_. otherwise this commit won't compile on its own which isn't a good thing.

File: app/src/main/java/mod/hilal/saif/activities/tools/AddCustomComponentActivity.java
Comment (JavkhlanK) [inline app/src/main/java/mod/hilal/saif/activities/tools/AddCustomComponentActivity.java:1]: a major UI change absolutely is not a refactoring btw, just admit it's a major feature lol

File: app/src/main/java/mod/hilal/saif/activities/tools/Tools.java
Comment (JavkhlanK) [inline app/src/main/java/mod/hilal/saif/activities/tools/Tools.java:1]: this shouldn't be a separate commit, make it part of the commit renaming the activities in both the manifest and code

File: app/src/main/java/mod/hilal/saif/components/ComponentsHandler.java [hunk 9]
```diff
@@ -574,4 +574,28 @@ private static ArrayList<HashMap<String, Object>> readCustomComponents() {
     public static void refreshCachedCustomComponents() {
         cachedCustomComponents = readCustomComponents();
     }
+
+    public static boolean isValidComponent(HashMap<String, Object> map) {
```
Comment (JavkhlanK) [inline app/src/main/java/mod/hilal/saif/components/ComponentsHandler.java:578]: when using Java collections, don't make unnecessary constraints like the parameter absolutely having to be a `HashMap`, and in this case, use the interface `Map` instead

File: app/src/main/java/mod/hilal/saif/components/ComponentsHandler.java [hunk 10]
```diff
@@ -574,4 +574,28 @@ private static ArrayList<HashMap<String, Object>> readCustomComponents() {
...
+        && map.containsKey("buildClass")
+        && map.containsKey("class")
+        && map.containsKey("description")
+        && map.containsKey("url")
+        && map.containsKey("additionalVar")
+        && map.containsKey("defineAdditionalVar")
+        && map.containsKey("imports");
+    }
+    
+    public static boolean isValidComponentList(ArrayList<HashMap<String, Object>> list) {
```
Comment (JavkhlanK) [inline app/src/main/java/mod/hilal/saif/components/ComponentsHandler.java:593]: same collections thing here, `List<Map<...` should work

**Review by JavkhlanK — CHANGES_REQUESTED** [review]
i haven't reviewed the major UI changes yet, but these changes are looking great!

when reading my comments, it might help to show files changed by commit and going through them in commit order.

**Review by JavkhlanK — APPROVED** [review]
other than the few changes that are needed, this looks both code-wise and in terms of UI really good!

please only implement the changes we discussed on Discord, I already have commits to implement the changes I just requested.

File: app/src/main/java/mod/elfilibustero/sketch/lib/ui/SketchFilePickerDialog.java [hunk 11]
```diff
@@ -0,0 +1,264 @@
...
+import java.util.Arrays;
+import java.util.List;
+
+import a.a.a.aB;
+import com.sketchware.remod.R;
+import mod.agus.jcoderz.lib.FileUtil;
+import mod.hey.studios.util.Helper;
+
+public class SketchFilePickerDialog extends aB {
+    private static Activity activity;
```
[the remark this answers is not in the document]
Comment (JavkhlanK) [reply app/src/main/java/mod/elfilibustero/sketch/lib/ui/SketchFilePickerDialog.java]: @elfilibusterismo that doesn't mean inner classes mustn't be static btw

File: app/src/main/java/mod/hilal/saif/activities/tools/AddCustomComponentActivity.java [hunk 12]
```diff
@@ -0,0 +1,258 @@
...
+            dialog.show();
+        } else {
+            HashMap<String, Object> map = list.get(0);
+            if (ComponentsHandler.isValidComponent(map)) {
+                setupViews(map);
+            } else {
+                SketchwareUtil.toastError(Helper.getResString(R.string.invalid_component));
+            }
+        }
+    }
```
[the remark this answers is not in the document]
Comment (JavkhlanK) [reply app/src/main/java/mod/hilal/saif/activities/tools/AddCustomComponentActivity.java:257]: @ilyassesalama how is this resolved already?

File: app/src/main/java/mod/elfilibustero/sketch/lib/ui/SketchFilePickerDialog.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (elfilibusterismo) [reply app/src/main/java/mod/elfilibustero/sketch/lib/ui/SketchFilePickerDialog.java]: Yes, I already changed it from my last commit.

File: app/src/main/java/mod/hilal/saif/activities/tools/AddCustomComponentActivity.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (JavkhlanK) [reply app/src/main/java/mod/hilal/saif/activities/tools/AddCustomComponentActivity.java]: this was slightly improved and I can't really come up with a better solution than creating a custom class extending `a.a.a.aB`
